# SQL Project

In this project, we are given a database of one of the competing startup companies which are developing new apps for book lovers. The database contains data about books, publishers, authors, as well as customer ratings and reviews of related books. 

## Objectives

Information extracted from the datasets will be used in making a price quotation for a new product.

## Dataset Check

In [1]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine

db_config = {'user': 'practicum_student',         # nama pengguna
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # kata sandi
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # port koneksi
             'db': 'data-analyst-final-project-db'}          # nama database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Books

In [2]:
# Displaying the first 10 rows of books dataset
books = pd.io.sql.read_sql('''SELECT * FROM books ''', con = engine)
books.head(10)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

Books dataset contains `book_id` as primary key and all information related to books including the identifier for author (`author_id`), title (`title`), number of pages (`num_pages`), date published (`publication_date`) and publisher identifier number (`publisher_id`). 

In [3]:
# Checking dataset information
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


Books dataset contains 1000 rows of data related to books.

In [4]:
# Checking earliest and latest publication date
print(books['publication_date'].min())
print(books['publication_date'].max())

1952-12-01
2020-03-31


The publication date in books dataset ranges from 1952-12-01 to 2020-03-31.

### Authors

In [5]:
# Displaying the first 10 rows of authors dataset
authors = pd.io.sql.read_sql('''SELECT * FROM authors ''', con = engine)
authors.head(10)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

In [6]:
# Checking the dataset information
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


Authors dataset contains 636 author names (`author`) and identifiers (`author_id`). The column `author_id` is the primary key of this dataset. Authors and books dataset has many-to-many relationship, meaning that one author can write more than one book, and one book can have more than one author.


### Publishers

In [7]:
# Displaying the first 10 rows of publishers dataset
publishers = pd.io.sql.read_sql('''SELECT * FROM publishers''', con = engine)
publishers.head(10)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

In [8]:
# Checking dataset information
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


Publisher dataset contains 340 rows of publishers name (`publisher`) and its identifier number (`publisher_id`). 

### Ratings

In [9]:
# Displaying the first 10 rows of ratings dataset
ratings = pd.io.sql.read_sql('''SELECT * FROM ratings''', con = engine)
ratings.head(10)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

In [10]:
# Checking dataset information
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


Ratings dataset contains all ratings (`rating`) given by users (`username`), the identifiers of the books rated (`book_id`) and the identifiers for each ratings (`rating_id`).

### Reviews

In [11]:
# Displaying the first 10 rows of reviews dataset
reviews = pd.io.sql.read_sql('''SELECT * FROM reviews''', con = engine)
reviews

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

In [12]:
# Checking dataset information
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


The dataset contains all reviews (`text`) given by the readers (`username`), along with the identifer number of books reviewed (`book_id`), and the primary key which is the identifier number of each review (`review_id`).

## Tasks

We will solve the following tasks:
* Count the number of books released after January 1, 2000.
* Calculate the number of user reviews and the average rating for each book.
* Identify publisher that has published the highest number of books, with more than 50 pages.
* Identify author with the highest average book rating: find books with a rating of at least 50.
* Calculate the average number of review texts among users who have rated more than 50 books.

### Books Released after January 1, 2000

In [13]:
# Finding the amount of books released after 2000-01-01
pd.io.sql.read_sql('''
SELECT COUNT(DISTINCT book_id) 
FROM books 
WHERE publication_date > CAST('2000-01-01' as Date)
''', con = engine)

count
0    819

There were 819 books in books dataset which were released after January 1, 2000.

### Number of Reviews and Average Rating for Each Book

In [14]:
# Finding the number of reviews and average rating for each book
pd.io.sql.read_sql('''
SELECT b.book_id, b.title, rev.reviews_count, rtg.avg_ratings
FROM books AS b
LEFT JOIN
(SELECT book_id, COUNT (DISTINCT review_id) AS reviews_count 
FROM reviews 
GROUP BY book_id) AS rev
ON b.book_id = rev.book_id
LEFT JOIN
(SELECT book_id, AVG(rating) as avg_ratings
FROM ratings
GROUP BY book_id) AS rtg
ON rev.book_id = rtg.book_id

''', con = engine)

book_id                                              title  \
0          1                                       'Salem's Lot   
1          2                 1 000 Places to See Before You Die   
2          3  13 Little Blue Envelopes (Little Blue Envelope...   
3          4  1491: New Revelations of the Americas Before C...   
4          5                                               1776   
..       ...                                                ...   
995      672   The Cat in the Hat and Other Dr. Seuss Favorites   
996       83    Anne Rice's The Vampire Lestat: A Graphic Novel   
997      221                          Essential Tales and Poems   
998      387                               Leonardo's Notebooks   
999      808                            The Natural Way to Draw   

     reviews_count  avg_ratings  
0              2.0     3.666667  
1              1.0     2.500000  
2              3.0     4.666667  
3              2.0     4.500000  
4              4.0     4.000000  
..             ...          ...  
995            NaN          NaN  
996            NaN          NaN  
997            NaN          NaN  
998            NaN          NaN  
999            NaN          NaN  

[1000 rows x 4 columns]

There were books without any reviews or ratings.

### Publisher with the Highest Book Count

In [15]:
# Finding publishers with the highest book count with num_pages >50
pd.io.sql.read_sql('''
SELECT sub.publisher
FROM (SELECT publisher, COUNT(DISTINCT book_id) as book_count 
FROM publishers INNER JOIN books ON publishers.publisher_id = books.publisher_id
WHERE num_pages > 50 
GROUP BY publisher
ORDER BY book_count DESC
LIMIT 1) AS sub''', con = engine)

publisher
0  Penguin Books

Penguin Books is the publisher with the highest number of books published.

### Author with the Highest Average Book Rating

In [18]:
# Find author with the highest average book rating: find books which have been rated for at least 50 times.
pd.io.sql.read_sql('''
SELECT SUB.author
FROM (SELECT author, AVG(rating) AS avg_rating 
FROM books 
INNER JOIN authors ON books.author_id = authors.author_id
INNER JOIN ratings ON books.book_id = ratings.book_id
GROUP BY author
HAVING COUNT(rating) >= 50 
ORDER BY avg_rating DESC
LIMIT 1) AS SUB''', con = engine)

author
0  Diana Gabaldon

Diana Gabaldon is the author with the highest average book rating.

### Number of Texts among Users with 50+ Books Rated

In [17]:
# Calculate the average number of texts among users who have rated more than 50 books.
pd.io.sql.read_sql('''
SELECT AVG(sub.text_count) as avg_text 
FROM (SELECT COUNT(text) as text_count 
FROM reviews 
WHERE username IN (SELECT username
FROM ratings
GROUP BY username
HAVING COUNT(rating_id) >50)
GROUP BY username) AS sub
''', con = engine)

avg_text
0  24.333333

Readers/users who rated more than 50 books typically give an average of 24 review texts.

## Conclusions

* Books dataset contains information of books published from 1952-12-01 to 2020-03-31.
* There were 819 books which were released after January 1, 2000 out of 1000 books in the dataset. Books dataset mostly contains books from the last two decades.
* There were books without any reviews or ratings.
* Penguin Books is the publisher with the highest number of books published.
* Diana Gabaldon is the author with the highest average book rating.
* Readers/users who rated more than 50 books typically give an average of 24 review texts.